persistent_queue/persistent_queue.py

```diff
@@ -0,0 +1,464 @@
+"""
+An implementation of a persistent queue. It is optimized for peeking at values
+and then deleting them off to top of the queue.
+"""
+
+import logging
+import os.path
+import pickle
+import shutil
+import struct
+import threading
+import time
+import uuid
+
+try:
+    import queue
+except ImportError:  # pragma: no cover
+    import Queue as queue
+
+LENGTH_STRUCT = 'I'
+HEADER_STRUCT = 'II'
+START_OFFSET = 4 + 4
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class PersistentQueue:
+    def __init__(self, maxsize=0, filename=None, dumps=pickle.dumps, loads=pickle.loads, flush_limit=1048576):
+        """
+        Creates a new PersistentQueue object and underlying file.
+
+        maxsize: upperbound limit of items that can be placed in the queue.
+        filename: must be a full path to the new file.
+        dumps: the function called for persisting queue items to the file.
+        loads: the function called for loading queue items from the file.
+        flush_limit: below this filesize, flush() is a no-op.
+        """
+        if filename is None:
+            raise ValueError("filename must be set, e.g. '/my/path/important.queue'")
```

Comment:
In my opinion, we don't need to check to make sure filename is a `str`. We don't do that for the rest of the parameters and `os.path.abspath(filename)` will fail if filename is an `int`. I don't feel strongly about this though, so if you want to add the check, go ahead. 

Here is what I think the signature should be:
```python
def __init__(self, filename, maxsize=0, dumps=pickle.dumps, loads=pickle.loads, flush_limit=1048576):
```